agent/xds/clusters.go

```diff
@@ -21,59 +26,146 @@ func clustersFromSnapshot(cfgSnap *proxycfg.ConfigSnapshot, token string) ([]pro
 	// Include the "app" cluster for the public listener
 	clusters := make([]proto.Message, len(cfgSnap.Proxy.Upstreams)+1)
 
-	clusters[0] = makeAppCluster(cfgSnap)
+	var err error
+	clusters[0], err = makeAppCluster(cfgSnap)
+	if err != nil {
+		return nil, err
+	}
 
 	for idx, upstream := range cfgSnap.Proxy.Upstreams {
-		clusters[idx+1] = makeUpstreamCluster(upstream.Identifier(), cfgSnap)
+		clusters[idx+1], err = makeUpstreamCluster(upstream, cfgSnap)
+		if err != nil {
+			return nil, err
+		}
 	}
 
 	return clusters, nil
 }
 
-func makeAppCluster(cfgSnap *proxycfg.ConfigSnapshot) *envoy.Cluster {
-	addr := cfgSnap.Proxy.LocalServiceAddress
-	if addr == "" {
-		addr = "127.0.0.1"
+func makeAppCluster(cfgSnap *proxycfg.ConfigSnapshot) (*envoy.Cluster, error) {
+	var c *envoy.Cluster
+	var err error
+
+	// If we have overriden local cluster config try to parse it into an Envoy cluster
+	if clusterJSONRaw, ok := cfgSnap.Proxy.Config["envoy_local_cluster_json"]; ok {
+		if clusterJSON, ok := clusterJSONRaw.(string); ok {
+			c, err = makeClusterFromUserConfig(clusterJSON)
+			if err != nil {
+				return c, err
+			}
+		}
 	}
-	return &envoy.Cluster{
-		Name: LocalAppClusterName,
-		// TODO(banks): make this configurable from the proxy config
-		ConnectTimeout: 5 * time.Second,
-		Type:           envoy.Cluster_STATIC,
-		// API v2 docs say hosts is deprecated and should use LoadAssignment as
-		// below.. but it doesn't work for tcp_proxy target for some reason.
-		Hosts: []*envoycore.Address{makeAddressPtr(addr, cfgSnap.Proxy.LocalServicePort)},
-		// LoadAssignment: &envoy.ClusterLoadAssignment{
-		//  ClusterName: LocalAppClusterName,
-		//  Endpoints: []endpoint.LocalityLbEndpoints{
-		//    {
-		//      LbEndpoints: []endpoint.LbEndpoint{
-		//        makeEndpoint(LocalAppClusterName,
-		//          addr,
-		//          cfgSnap.Proxy.LocalServicePort),
-		//      },
-		//    },
-		//  },
-		// },
+
+	if c == nil {
```

Comment:
nit: I'd probably just `return c, nil` in the If above rather than have the rest conditional on the first part - slightly easier to read. That said you probably copied the pattern from my code else where 😄

Reply:
Ah yeah the upstream below shows why - in other cases we need to always inject TLS stuff even if we got the config from the user so we continue processing. Leaving this consistent is fine.